db.js

```diff
@@ -1,44 +1,71 @@
-import sqlite from 'sqlite3';
+import sqlite3 from 'sqlite3';
+import { open } from 'sqlite'
 
 const TABLE_NAME = 'cfgov';
 
 class DB {
-  constructor (dbFilePath) {
-    this.db = new sqlite.Database(dbFilePath, (err) => {
-      if (err) {
-        console.error('Could not connect to database', err);
-      }
+  constructor (db) {
+    this.db = db;
+  }
+
+  static async connect (dbFilePath) {
+    const db = await open({
+      filename: dbFilePath,
+      driver: sqlite3.Database
     });
+
+    return new DB( db );
   }
 
-  createTable () {
-    const sql = `
+  async createTables () {
+    await this.db.run(`
       CREATE TABLE IF NOT EXISTS ${TABLE_NAME} (
-        id PRIMARY KEY,
-        path TEXT,
+        path TEXT PRIMARY KEY,
         title TEXT,
         components TEXT,
         links TEXT,
         pageHash TEXT,
         pageHtml TEXT,
         timestamp TEXT
-      )`;
-    return this.run(sql);
+      )
+    `);
+
+    await this.db.run(`
+      CREATE VIRTUAL TABLE IF NOT EXISTS ${TABLE_NAME}_fts USING fts5(
+        path,
+        pageHtml,
+        content=${TABLE_NAME}
+      )
+    `);
+
+    await this.db.run(`
+      CREATE TRIGGER IF NOT EXISTS insert_${TABLE_NAME}_fts
+      AFTER INSERT ON ${TABLE_NAME}
+      BEGIN
+        INSERT INTO ${TABLE_NAME}_fts(
+          path,
+          pageHtml
+        ) VALUES (
+          new.path,
+          new.pageHtml
+        );
+      END
+    `);
   }
 
-  insert (record) {
+  async insert (record) {
     const sql = `
-      INSERT OR REPLACE INTO ${TABLE_NAME} (
-        id,
+      INSERT INTO ${TABLE_NAME} (
```

Comment:
I get a `[Error: SQLITE_CONSTRAINT: UNIQUE constraint failed: cfgov.path]` error when running the script against a DB that already exists

```suggestion
      INSERT OR REPLACE INTO ${TABLE_NAME} (
```

Reply:
@contolini unfortunately the [external content tables](https://www.sqlite.org/fts5.html#external_content_tables) feature I propose here doesn't support `REPLACE`.

I was wondering whether we want to support running against a DB that already exists. I don't necessarily see the value in that particular use case - if you have a crawl from time stored in the DB, and then restart a new crawl at some other time, until the crawl is finished the state of the DB isn't particularly useful since it's a mix of the two times. Would it be better to constrain a DB to a single crawl? We could add some functionality to the script that exits if the file exists or recreates it if so desired (via some kind of `-f, --force` option).

For the purposes of a frontend client, I had been thinking we'd want to run the crawl against some DB location X and then move it to location Y once finished for purposes of serving; i.e. the client wouldn't be trying to serve a DB file that was currently being exposed via the client. What do you think?

Reply:
> @contolini unfortunately the external content tables feature I propose here doesn't support REPLACE.

That would only be an issue with the FTS table insertion on L45, though, right?

Regardless, I like your recommendations. I'm cool with trashing the DB at the start of the script and always starting from scratch. When the crawl is done, we can timestamp it (e.g. `20220404-cfgov-crawl.sqlite3`) and keep daily/weekly/monthly snapshots for future querying purposes. The latest crawl can also be copied to a file called `latest.sqlite3` or something and the app reads from that file.